### DISCUSSION TITLE
XMLParserのエラーが発生し、解析に失敗

### DISCUSSION THREAD

**Discussion by oyui7272** [lit-iphone-course/lit-iphone-course#25] [Question-original] [answered]
**(1) 実現したいこと**
responseDataに入っている文字を解析して、必要なデータを取り出したい

**(2)  実現するために自分で試した内容と結果**
https://softmoco.com/swift/swift-how-to-parse-xml.php
こちらの記事を参考に、XMLParserを使った解析の処理をするコードを書きました。
実行前にxcode上でエラーは出ないのですが、実行した結果、「NSXMLParserErrorDomain error 64.」という理由で、解析できません。

**(3) 調べた内容や資料**
https://stackoverflow.com/questions/18442450/nsxmlparsererrordomain-error-64
古い記事ですが、解析できない文字が含まれる、ということかと思っています。

https://www.w3schools.com/xml/xml_validator.asp
こちらのサイトを使って、XMLの構文チェックをしたところ、構文自体にはエラーはなさそうでした。

**(4) 原因だと思われそうな箇所**
あまりわかっていないのですが、解析方法を記述した671行目 ~ 716行目あたりが原因だと思います。

**(5) Githubのリンク**
https://github.com/yuipi7272/NanoTour.git

**(6) その他**
・API参考URL
https://developer.yahoo.co.jp/webapi/map/openlocalplatform/v1/localsearch.html#start

・ 取り出したいデータ
671行目 ~ 716行目あたりに記述しています
・場所名 Name
・カテゴリー Genre → Name
・住所 Address
・緯度・経度 Coordinates

**Comment by yukiny0811** [comment, marked as answer]
XMLは <?xml... の前に空白があるといけないから、画像みたいに1行目のインデントを消してあげればなおるよー！

これだとエラー↓
```.xml
    <?xml........
```

これだと大丈夫↓
```.xml
 <?xml........
```

<img width="197" alt="image" src="https://user-images.githubusercontent.com/28947703/190851632-0bb540f2-1594-455b-943e-84a867c1df6e.png">

**Reply by oyui7272** [reply]
ありがとうございます！！！そこだったのですね😳😳😳
動きました！

全然気にしていなかった部分だったのでびっくりしました...笑

**Comment by txxxxc** [comment]
https://map.yahooapis.jp/search/local/V1/localSearch?appid=自分のAPIKey&gc=0115001&output=json 
でJSONが返ってくるからxmlパーサ使うよりもjsonを取得してデコードしたほうが良さそうです！！
(パラメータに`output=json`がついているとjsonが返ってくるみたいです...!)

[【Swift】JSONデータをデコードする方法！JSONDecoderクラスの使い方](https://tech.amefure.com/swift-json-decoder)

**Reply by txxxxc** [reply]
多分メンバーに教えることを踏まえるとSwiftを使ってjson変換する方法を知ってもらいたい気持ちはあります...!
(xml返すAPIよりjson返すAPIのほうが多いので...!)
でも結構頑張って調べてくれてるっぽいので、xmlで行けそうだったらxmlでOKです！！

**Reply by yukiny0811** [reply]
たしかにJsonの方がメンバーに教える時は使うね！
まぁでもXML解析できればJsonもいけるっしょ！

**Reply by oyui7272** [reply]
JSONに変形できるの気づいてなかったです😳
ありがとうございます！

Day4後になるかもしれないですが、JSONで取得するようにも修正しておきます！

**Comment by oyui7272** [comment]
何度もすみません！追加でこちらで質問しても大丈夫でしょうか...??

APIから直接XMLParseを使って解析するように修正したのですが、
744行目(callYOLPAPI関数内)でdumpを書いた場合はplace_arrayの内容が出力されるのですが、
622行目(viewDidLoad関数内)でdumpを書いた場合はplace_array内の内容が出力されず。

viewDidLoad内でもplace_array内の内容を表示させるためにはどのようにすればよいのでしょうか？

原因として、741行目 ~ 743行目parserを定義するときもselfをつけないとエラーが出てしまったため、近い理由があるのかと考えているのですが、分からないため、教えていただけると嬉しいです🙏

**Reply by txxxxc** [reply]
今手元にパソコンなくて携帯で見てるので、間違ってたらごめんだけど、プログラムが非同期で実行されているのが、原因だと思います！

**Reply by txxxxc** [reply]
requrst.responseString（700行あたりのやつ）はAPI通信が終わった後つまり、レスポンスが帰ってきた後に実行されますが、600行当たりのやつはAPIが返ってくる前に実行されるので、空が表示されるという流れです！

**Reply by oyui7272** [reply]
ありがとうございます！
なるほど...!!

request.responseStringはcallYOLPAPI関数の中で処理しているのですが、
viewDidLoad関数の中での処理の順番として、
1. callYOLPAPI関数を呼ぶ
2. callYOLPAPI関数内の処理が終わっていなくても、次のdump関数が呼ばれる
という認識であっているでしょうか...??

**Reply by txxxxc** [reply]
そういうことです！
多分それが原因かなとおもってますー！

**Reply by oyui7272** [reply]
ありがとうございます！！なるほど...！分かりました！！そんなこともあるのですね...😳✨

**Reply by txxxxc** [reply]
非同期処理とかで調べてもらうとわかるかも...!

**Reply by oyui7272** [reply]
ありがとうございます！調べてみます！！
